### ISSUE TITLE
[BUG] Wordpress helm chart in RWX mode has permission issues

### ISSUE THREAD

**Issue by linucksrox** [longhorn/longhorn#3661]
## Describe the bug

When deploying the bitnami Wordpress helm chart in RWO mode, everything works correctly. But when changing to RWX mode, I get to the white screen of death in Wordpress. After logging into the admin console I can see that the theme is not installed, which leads me to check the file system permissions in the pod.

File system permissions and ownership in RWO mode (working configuration):
```
$ ls -alh /bitnami/wordpress/wp-content/
total 32K
drwxrwxr-x  7 1001 1001 4.0K Feb 24 14:04 .
drwxrwsr-x  3 root 1001 4.0K Feb 24 14:04 ..
-rw-rw-r--  1 1001 1001   28 Feb 24 14:04 index.php
drwxrwxr-x  2 1001 1001 4.0K Feb 24 14:04 languages
drwxrwxr-x 12 1001 1001 4.0K Feb 24 14:04 plugins
drwxrwxr-x  5 1001 1001 4.0K Feb 24 14:04 themes
drwxrwxr-x  2 1001 1001 4.0K Feb 24 14:04 upgrade
drwxrwxr-x  3 1001 1001 4.0K Feb 24 14:04 uploads
```

File system permissions and ownership in RWX mode:
```
$ ls -alh bitnami/wordpress/wp-content/
total 24K
drwxr-xr-x 5 1001 root 4.0K Feb 24 14:28 .
drwxrwxrwx 3 root root 4.0K Feb 24 14:24 ..
-rw-r--r-- 1 1001 root   28 Feb 24 14:24 index.php
drwxr-xr-x 2 1001 root 4.0K Feb 24 14:24 languages
drwxr-xr-x 8 1001 root 4.0K Feb 24 14:26 plugins
drwxr-xr-x 3 1001 root 4.0K Feb 24 14:28 uploads
```

So in conclusion, it looks like RWX mode is causing some permission issues for the pod, which I suspect is related to the NFS share being used in that case.

## Environment

 - Longhorn version: 1.2.3
 - Installation method (e.g. Rancher Catalog App/Helm/Kubectl): kubectl
 - Kubernetes distro (e.g. RKE/K3s/EKS/OpenShift) and version: kubeadm
   - Number of management node in the cluster: 3
   - Number of worker node in the cluster: 6
 - Node config
   - OS type and version: Ubuntu 20.04
   - CPU per node: 4
   - Memory per node: 4GB
   - Disk type(e.g. SSD/NVMe): NVMe and mechanical
   - Network bandwidth between the nodes: virtual
 - Underlying Infrastructure (e.g. on AWS/GCE, EKS/GKE, VMWare/KVM, Baremetal): Proxmox (KVM)
 - Number of Longhorn volumes in the cluster: 2

**Comment by innobead** [comment]
ref: https://github.com/longhorn/longhorn/issues/2418 cc @c3y1huang

**Comment by derekbit** [comment]
@linucksrox 
Can you provide the manifest or step for reproducing the permission error?

**Comment by linucksrox** [comment]
@derekbit Sure, I just ran `helm install bitnami/wordpress -f values.yaml --version=13.0.15` with the following yaml file (just with a different ingress hostname)
[values.yaml.txt](https://github.com/longhorn/longhorn/files/8141742/values.yaml.txt)

**Comment by jenting** [comment]
It's a limitation because of the Longhorn design on the RWX volume. Ref to https://github.com/longhorn/longhorn/issues/2418#issuecomment-809925321

**Comment by linucksrox** [comment]
@jenting If I'm reading this correctly, that issue you referred to was resolved by setting permissions to 0777 on NFS shares, see here: https://github.com/longhorn/longhorn-share-manager/pull/12
The milestone for that was v1.1.1 and I'm running v1.2.3 so I don't understand why I would still be seeing this issue if the permissions have been changed.

**Comment by joshimoo** [comment]
For RWX volumes kubernetes does not do recursive chown/chmod's as it does for RWO volumes.
Can you clarify the expected permissions / ownership model, that the wordpress installation requires?

**Comment by linucksrox** [comment]
@joshimoo I'm probably misunderstanding your question, but I think the expected permissions should be the first model listed in my original post. So the owner/group should be 1001/1001 and permissions are 775/664 for folders/files.
```
$ ls -alh /bitnami/wordpress/wp-content/
total 32K
drwxrwxr-x  7 1001 1001 4.0K Feb 24 14:04 .
drwxrwsr-x  3 root 1001 4.0K Feb 24 14:04 ..
-rw-rw-r--  1 1001 1001   28 Feb 24 14:04 index.php
drwxrwxr-x  2 1001 1001 4.0K Feb 24 14:04 languages
drwxrwxr-x 12 1001 1001 4.0K Feb 24 14:04 plugins
drwxrwxr-x  5 1001 1001 4.0K Feb 24 14:04 themes
drwxrwxr-x  2 1001 1001 4.0K Feb 24 14:04 upgrade
drwxrwxr-x  3 1001 1001 4.0K Feb 24 14:04 uploads
```

I'm a little confused as to what I'm being told here... When I opened this issue, I was referred to issue #2418 which has been resolved. The Wordpress example is just something I was using to evaluate RWX volumes on Longhorn because Wordpress is a pretty straightforward thing to install and test.

**Comment by derekbit** [comment]
@linucksrox 
I tried to create the wordpress by `helm install wordpress  bitnami/wordpress -f values.yaml` using longhorn RWX and RWO volumes. I noticed that folders with permission 755 were created firstly and were updated to 775 gradually in both cases.

The `fsGroup` in `values.yaml` is already set to `1001`, but the `group` of the files is still `root` rather than `1001`. This might be related to the `fsGroupPolicy=ReadWriteOnceWithFSType` setting in the `driver.longhorn.io`.

Ref: https://kubernetes-csi.github.io/docs/support-fsgroup.html

**Comment by linucksrox** [comment]
Thank you for taking more time to look at this. I don't have enough experience with Kubernetes to really understand fsGroupPolicy but I get the general idea. So is this something that would need to change on Longhorn, or something I need to modify in the values.yaml file for the helm chart?

**Comment by derekbit** [comment]
`values.yaml` looks fine.
The `CSIDriver` creation is embedded in longhorn-manager now, so we need to do more investigation on `fsGroupPolicy`.

cc @joshimoo @jenting @innobead

**Comment by joshimoo** [comment]
cc @innobead For the fsGroupPolicy specs on the driver, we can revisit once we have done a min kubernetes version update for v1.3
https://kubernetes-csi.github.io/docs/support-fsgroup.html

Remember there was a couple of other issues, that we can address when doing the min version update, i.e. go version bump, client-go bump.

**Comment by joshimoo** [comment]
In regard to "fsGroupPolicy" by default kubernetes defaulted to `ReadWriteOnceWithFSType` which implied that no permission/ownership changes would be performed as part of kubelet for RWX volumes. 

@linucksrox as a workaround, you can run an init container with the RWX workload to do the `chmod+chown`

**Comment by innobead** [comment]
YES, thanks @joshimoo and @derekbit analysis.

Right now, we are using `release-1.18` code base, so this is why we don't have any custom behavior for `ReadWriteOnceWithFSType ` of CSIDriverSpec and it is just the default behavior of `fsGroupPolicy ` (`ReadWriteOnceWithFSType `) instead as @joshimoo and @derekbit mentioned.

The workaround is as @joshimoo mentioned, and the end goal is to define the expected behavior by providing the setting for `fsGroupPolicy`. Also, we can take a look if we would like to leave volume fs permission to our CSI handle completely instead of relying on kubelet (https://kubernetes-csi.github.io/docs/support-fsgroup.html#delegate-fsgroup-to-csi-driver). 

ref:
- https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/storage/v1/types.go
- https://github.com/kubernetes/kubernetes/blob/release-1.18/staging/src/k8s.io/api/storage/v1/types.go

**Comment by PhanLe1010** [comment]
The feature https://kubernetes-csi.github.io/docs/support-fsgroup.html#delegate-fsgroup-to-csi-driver seems to have to wait since it is only available in beta in Kubernetes v1.23. 

While waiting, we should check setting `fsGroupPolicy=File` to see if it can meet our requirement: respect the setting `pod.spec.securityContext.fsGroupChangePolicy` for both RWO and RWX which is:
* if `pod.spec.securityContext.fsGroupChangePolicy=OnRootMismatch` don't recursive change the volume ownership and permissions unless the root is mismatch. (if it doesn't meet this requirement, there will be a big delay when mounting volume, see more at https://github.com/longhorn/longhorn/issues/2131#issuecomment-778897129)
* if `pod.spec.securityContext.fsGroupChangePolicy=always` always change the volume ownership.

**Comment by baskinsy** [comment]
I have the same issue on a 1.23.7 RKE2 cluster with longhorn 1.2.4. Wordpress still fails to write to the RWX volume due to permissions. Any workaround? Thank you.

**Comment by tk2023** [comment]
+1 on the issue.

I am trying to install bitnami/wordpress package version 19.0.4 and it still have permission issue as:

`wordpress 06:09:08.45 INFO  ==> Persisting WordPress installation
cp: cannot create regular file '/bitnami/wordpress/wp-config.php': Permission denied`

**Comment by yaleman** [comment]
not stale!

**Comment by HyveBB** [comment]
Other than the init container is there anyway around this in longhorn?

**Comment by yaleman** [comment]
not stale! 🥳
